File: app-kotlin/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/data/utilities/geometry/Order0.java
/*
 * Copyright (c) 1998, 2018, Oracle and/or its affiliates. All rights reserved.
 * ORACLE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 */

package com.TritiumGaming.phasmophobiaevidencepicker.data.utilities.geometry;

import androidx.annotation.NonNull;

/** @noinspection CanBeFinal, CanBeFinal */
final class Order0 extends Curve {
    private final double x;
    private final double y;

    public Order0(double x, double y) {
        super(INCREASING);
        this.x = x;
        this.y = y;
    }

    public int getOrder() {
        return 0;
    }

    public double getXTop() {
        return x;
    }

    public double getYTop() {
        return y;
    }

    public double getXBot() {
        return x;
    }

    public double getYBot() {
        return y;
    }

    public double getXMin() {
        return x;
    }

    public double getXMax() {
        return x;
    }

    public double getX0() {
        return x;
    }

    public double getY0() {
        return y;
    }

    public double getX1() {
        return x;
    }

    public double getY1() {
        return y;
    }

    public double XforY(double y) {
        return y;
    }

    public double TforY(double y) {
        return 0;
    }

    public double XforT(double t) {
        return x;
    }

    public double YforT(double t) {
        return y;
    }

    public double dXforT(double t, int deriv) {
        return 0;
    }

    public double dYforT(double t, int deriv) {
        return 0;
    }

    public double nextVertical(double t0, double t1) {
        return t1;
    }

    public int crossingsFor(double x, double y) {
        return 0;
    }

    public boolean accumulateCrossings(@NonNull Crossings c) {
        return (x > c.getXLo() &&
                x < c.getXHi() &&
                y > c.getYLo() &&
                y < c.getYHi());
    }

    public void enlarge(@NonNull Rectangle2D r) {
        r.add(x, y);
    }

    @NonNull
    public Curve getSubCurve(double ystart, double yend, int dir) {
        return this;
    }

    @NonNull
    public Curve getReversedCurve() {
        return this;
    }

    public int getSegment(@NonNull double[] coords) {
        coords[0] = x;
        coords[1] = y;
        return PathIterator.SEG_MOVETO;
    }
}

File: core-common/src/main/kotlin/com/tritiumgaming/core/common/graphics/geometry/Order0.kt
/*
 * Copyright (c) 1998, 2018, Oracle and/or its affiliates. All rights reserved.
 * ORACLE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 *
 */
package com.tritiumgaming.core.common.graphics.geometry

/**
 * @noinspection CanBeFinal, CanBeFinal
 */
internal class Order0(
    override val x1: Double,
    override val y1: Double
) : Curve(INCREASING) {
    override val order: Int
        get() = 0

    override val xTop: Double = 0.0
    override val yTop: Double = 0.0
    override val xBot: Double = 0.0
    override val yBot: Double = 0.0
    override val xMin: Double = 0.0
    override val xMax: Double = 0.0
    override val x0: Double = 0.0
    override val y0: Double = 0.0

    override val reversedCurve: Curve
        get() = this

    override fun XforY(y: Double): Double { return y }

    override fun TforY(y: Double): Double { return 0.0 }

    override fun XforT(t: Double): Double { return x1 }

    override fun YforT(t: Double): Double { return y1 }

    override fun dXforT(t: Double, deriv: Int): Double { return 0.0 }

    override fun dYforT(t: Double, deriv: Int): Double { return 0.0 }

    override fun nextVertical(t0: Double, t1: Double): Double { return t1 }

    override fun crossingsFor(x: Double, y: Double): Int { return 0 }

    override fun accumulateCrossings(c: Crossings): Boolean {
        return (x1 > c.xLo && x1 < c.xHi && y1 > c.yLo && y1 < c.yHi)
    }

    override fun enlarge(r: Rectangle2D?) {
        r?.add(x1, y1)
    }

    override fun getSubCurve(ystart: Double, yend: Double, dir: Int): Curve { return this }

    override fun getSegment(coords: DoubleArray?): Int {
        coords?.set(0, x1)
        coords?.set(1, y1)
        return PathIterator.SEG_MOVETO
    }
}
